src/main/java/hexgui/gui/HexGui.java

```diff
@@ -0,0 +1,2393 @@
+
+    Vector<Pair<HexColor, HexPoint>> vr = StringUtils.parseVariation(fx.substring(var + 3, label));
+    for (int i = 0; i < vr.size(); ++i) {
+      m_guiboard.setColor(vr.get(i).second, vr.get(i).first);
+      m_guiboard.setText(vr.get(i).second, Integer.toString(i + 1));
+    }
+
+    String label_str = fx.substring(label + 5, text).trim();
+    showInferiorCells(label_str);
+
+    String prog_str = fx.substring(text + 4).trim();
+    String[] levels = prog_str.split(" ");
+
+    double contribution = 1.0;
+    double progress = 0.0;
+    for (int i = 0; i < levels.length; ++i) {
+      String[] nums = levels[i].trim().split("/");
+      int cur = Integer.decode(nums[0]).intValue();
+      int max = Integer.decode(nums[1]).intValue();
+      progress += contribution * cur / max;
+      contribution *= 1.0 / max;
+    }
+    m_guiboard.repaint();
+    m_statusbar.setMessage(fx.substring(text + 5));
+    m_statusbar.setProgress(progress);
+  }
+
+  private void guifx_dfpn(String fx) {
+    m_guiboard.clearMarks();
+    m_guiboard.aboutToDirtyStones();
+
+    int var = fx.indexOf("VAR");
+    int label = fx.indexOf("LABEL");
+    int text = fx.indexOf("TEXT");
+
+    Vector<Pair<HexColor, HexPoint>> vr = StringUtils.parseVariation(fx.substring(var + 3, label));
+    for (int i = 0; i < vr.size(); ++i) {
+      m_guiboard.setColor(vr.get(i).second, vr.get(i).first);
+      m_guiboard.setText(vr.get(i).second, Integer.toString(i + 1));
+      m_guiboard.setAlphaColor(vr.get(i).second, Color.blue);
+    }
+    String label_str = fx.substring(label + 5, text).trim();
+    showDfpnBounds(label_str);
+
+    m_guiboard.repaint();
+    m_statusbar.setMessage(fx.substring(text + 5));
+  }
+
+  private void showDfpnBounds(String str) {
+    Vector<Pair<String, String>> pairs = StringUtils.parseStringPairList(str);
+    for (int i = 0; i < pairs.size(); i++) {
+      HexPoint point = HexPoint.get(pairs.get(i).first);
+      String value = pairs.get(i).second;
+      m_guiboard.setText(point, value);
+      if (value.trim().equals("W")) m_guiboard.setAlphaColor(point, Color.green);
+      else if (value.trim().equals("L")) m_guiboard.setAlphaColor(point, Color.red);
+    }
+  }
+
+  /** Draws the inferior cells to the gui board. */
+  private void showInferiorCells(String str) {
+    Vector<Pair<String, String>> pairs = StringUtils.parseStringPairList(str);
+    for (int i = 0; i < pairs.size(); i++) {
+      HexPoint point = HexPoint.get(pairs.get(i).first);
+      String value = pairs.get(i).second;
+
+      if (value.charAt(0) == 'f') // fill-in
+      {
+        assert (3 == value.length());
+        if (value.charAt(1) == 'd') // dead
+        m_guiboard.setAlphaColor(point, Color.cyan);
+        else if (value.charAt(1) == 'p') // permanently inferior
+        m_guiboard.setAlphaColor(point, Color.gray);
+        else // captured
+        {
+          assert (value.charAt(1) == 'c');
+          m_guiboard.setAlphaColor(point, Color.red);
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1585:1593, 1602:1610.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.